### ISSUE TITLE
`NullPointerException` inside `asMemberOf`

### ISSUE THREAD

**Issue by laech** [mapstruct/mapstruct#2683]
I'm getting the following error when using v1.5.0-Beta1 and compiling with Gradle (7.3.1), it happens randomly so I'm not sure how to recreate it reliably:

```
error: Internal error in the mapping processor: java.lang.NullPointerException: Cannot invoke "com.sun.tools.javac.code.Symbol.getEnclosingElement()" because "sym" is null
at jdk.compiler/com.sun.tools.javac.model.JavacTypes.asMemberOf(JavacTypes.java:298)
at org.mapstruct.ap.internal.util.AbstractTypeUtilsDecorator.asMemberOf(AbstractTypeUtilsDecorator.java:129)
at org.mapstruct.ap.internal.util.Filters.getWithinContext(Filters.java:147)
at org.mapstruct.ap.internal.util.Filters.getReturnType(Filters.java:118)
at org.mapstruct.ap.internal.util.Filters.recordAccessorsIn(Filters.java:104)
at org.mapstruct.ap.internal.model.common.Type.getPropertyReadAccessors(Type.java:593)
at org.mapstruct.ap.internal.model.common.Type.getAlternativeTargetAccessors(Type.java:882)
at org.mapstruct.ap.internal.model.common.Type.getPropertyWriteAccessors(Type.java:643)
at org.mapstruct.ap.internal.model.BeanMappingMethod$Builder.build(BeanMappingMethod.java:209)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMappingMethods(MapperCreationProcessor.java:416)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.getMapper(MapperCreationProcessor.java:179)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:133)
at org.mapstruct.ap.internal.processor.MapperCreationProcessor.process(MapperCreationProcessor.java:81)
at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:338)
at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:318)
at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:267)
at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:166)
at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
at org.gradle.api.internal.tasks.compile.processing.IsolatingProcessor.process(IsolatingProcessor.java:50)
at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.access$401(TimeTrackingProcessor.java:37)
at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:99)
at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor$5.create(TimeTrackingProcessor.java:96)
at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.track(TimeTrackingProcessor.java:117)
at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.process(TimeTrackingProcessor.java:96)
at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1025)
at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:940)
at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1269)
at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1384)
at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1261)
at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:935)
at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:89)
at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:567)
at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)
at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by Zegveld** [comment]
From the stacktrace I can determine that this takes place when working with records.
MapStruct tries to retrieve the get-method for the record field, but this seems to fail and return a `null` value where we always expect it to succeed.

I did see some changes to the way we deal with records in 1.5.0.Beta2, but I do not expect that this has fixed it.
@laech, If you could determine what record setup causes this issue, that would help.

Perhaps a check could be added at that location in the code to give a compiler error when this situation occurs. Something like: `Unexpected situation found while processing record <record detail information>. Please report this at the MapStruct Issue Tracker under issue 2683.`
What do you think, @filiphr ?

**Comment by laech** [comment]
@Zegveld I did a bit more digging on this, seems this happens randomly when I host the project inside WSL2 under Windows. Under native Linux I haven't experience this problem.

I'm not able to reproduce the issue reliably, but it did happen once today for a mapper that looks like this:

```java
@Mapper(
    componentModel = "spring",
    unmappedTargetPolicy = ReportingPolicy.IGNORE,
    injectionStrategy = InjectionStrategy.CONSTRUCTOR)
public interface ApiProfilePendingExpirationEventMapper {

  MyRecord toRecord(MyClass myClass);
}

public record MyRecord(String myA, String myB) {}

@lombok.Value
public class MyClass {
  String myA;
  String myB;
}

```

**Comment by laech** [comment]
Just ran into this under Linux as well.

**Comment by filiphr** [comment]
Looking at the stacktrace this seems to be an identical problem to #2294 and #2374. There was a bug in JDK 15 and JDK 16 that was causing `RecordComponentElement#getAccessor` to return a `null`. This issue was tracked in #2375 and was fixed in Java 17 with the issue [JDK-8258535](https://bugs.openjdk.java.net/browse/JDK-8258535).

@laech can you please let me know which Java you were using when this happened to you?

You are most likely seeing it randomly due to the Gradle incremental compilation. Most likely you are seeing it when you do not have any changes in your records but you do have in the mappers. Gradle then invokes javac only with the mapper and the records are available through their bytecode, i.e. they are like in another module.

@Zegveld I understand your point, we could have a null check after the `RecordComponentElement#getAccessor` call and return a compile error telling the user to upgrade to Java 17 due to the JDK bug 😉 . However, I am not sure that it is worth the effort. JDK 17 is currently the LTS version and I do not believe that there is a vendor that is still supporting JDK 16 (which is the first version that has records as a non preview feature).

**Comment by laech** [comment]
Thanks @filiphr, the project was indeed using JDK 16, happy to close this.
